src/prefect/new_flow_engine.py

```diff
@@ -0,0 +1,305 @@
+        result_factory = getattr(FlowRunContext.get(), "result_factory", None)
+        terminal_state = await return_value_to_state(
+            await resolve_futures_to_states(result),
+            result_factory=result_factory,
+        )
+        await self.set_state(terminal_state)
+        return result
+
+    async def handle_exception(
+        self,
+        exc: Exception,
+        msg: Optional[str] = None,
+        result_factory: Optional[ResultFactory] = None,
+    ) -> State:
+        context = FlowRunContext.get()
+        state = await exception_to_failed_state(
+            exc,
+            message=msg or "Flow run encountered an exception:",
+            result_factory=result_factory or getattr(context, "result_factory", None),
+        )
+        state = await self.set_state(state)
+        if self.state.is_scheduled():
+            state = await self.set_state(Running())
+        return state
+
+    async def create_subflow_task_run(
+        self, client: PrefectClient, context: FlowRunContext
+    ) -> TaskRun:
+        dummy_task = Task(
+            name=self.flow.name, fn=self.flow.fn, version=self.flow.version
+        )
+        task_inputs = {
+            k: await collect_task_run_inputs(v) for k, v in self.parameters.items()
+        }
+        parent_task_run = await client.create_task_run(
+            task=dummy_task,
+            flow_run_id=(
+                context.flow_run.id if getattr(context, "flow_run", None) else None
+            ),
+            dynamic_key=_dynamic_key_for_task_run(context, dummy_task),
+            task_inputs=task_inputs,
+            state=Pending(),
+        )
+        return parent_task_run
+
+    async def get_most_recent_flow_run_for_parent_task_run(
+        self, client: PrefectClient, parent_task_run: TaskRun
+    ) -> "Union[FlowRun, None]":
+        """
+        Get the most recent flow run associated with the provided parent task run.
+
+        Args:
+            - An orchestration client
+            - The parent task run to get the most recent flow run for
+
+        Returns:
+            The most recent flow run associated with the parent task run or `None` if
+            no flow runs are found
+        """
+        flow_runs = await client.read_flow_runs(
+            flow_run_filter=FlowRunFilter(
+                parent_task_run_id={"any_": [parent_task_run.id]}
+            ),
+            sort=FlowRunSort.EXPECTED_START_TIME_ASC,
+        )
+        return flow_runs[-1] if flow_runs else None
+
+    async def create_flow_run(self, client: PrefectClient) -> FlowRun:
+        flow_run_ctx = FlowRunContext.get()
+
+        parent_task_run = None
+        # this is a subflow run
+        if flow_run_ctx:
+            parent_task_run = await self.create_subflow_task_run(
+                client=client, context=flow_run_ctx
+            )
+            # If the parent task run already completed, return the last flow run
```

Comment:
Is there a reason we couldn't use idempotency keys here?

Reply:
@desertaxle correct me if i'm wrong but this copies the current implementation verbatim, right?